Source/Plugins/NetworkEvents/NetworkEvents.cpp

```diff
@@ -35,200 +35,75 @@ const int MAX_MESSAGE_LENGTH = 64000;
-    return Svec;
-}
-
-
-StringTS::StringTS (MidiMessage& event)
-{
-    const uint8* dataptr = event.getRawData();
-    const int bufferSize = event.getRawDataSize();
-    len = bufferSize - 6 - 8; // -6 for initial event prefix, -8 for timestamp at the end
-
-    memcpy (&timestamp, dataptr + 6 + len, 8); // remember to skip first six bytes
-    str = new uint8[len];
-    memcpy (str,dataptr + 6, len);
-}
-
-
-StringTS& StringTS::operator= (const StringTS& rhs)
-{
-    delete (str);
-    len = rhs.len;
-    str = new uint8[len];
-    memcpy (str,rhs.str,len);
-    timestamp = rhs.timestamp;
-
-    return *this;
-}
-
-
-String StringTS::getString()
-{
-    return String ((const char*)str,len);
-}
-
-
-StringTS::StringTS (String S)
-{
-    Time t;
-    str = new uint8[S.length()];
-    memcpy (str, S.toRawUTF8(), S.length());
-    timestamp = t.getHighResolutionTicks();
-
-    len = S.length();
-}
-
-
-StringTS::StringTS (String S, int64 ts_software)
-{
-    str = new uint8[S.length()];
-    memcpy (str, S.toRawUTF8(), S.length());
-    timestamp = ts_software;
-
-    len = S.length();
-}
-
-
-StringTS::StringTS (const StringTS& s)
-{
-    str = new uint8[s.len];
-    memcpy (str, s.str, s.len);
-    timestamp = s.timestamp;
-    len = s.len;
-}
-
-
-StringTS::StringTS (unsigned char* buf, int _len, int64 ts_software) 
-    : len       (_len)
-    ,timestamp  (ts_software)
-{
-    str = new juce::uint8[len];
-    for (int k = 0; k < len; ++k)
-        str[k] = buf[k];
-}
-
-
-StringTS::~StringTS()
-{
-    delete str;
-}
-
-
-/*********************************************/
-void* NetworkEvents::zmqcontext = nullptr;
+NetworkEvents::ZMQContext* NetworkEvents::sharedContext = nullptr;
+CriticalSection NetworkEvents::sharedContextLock{};
 
 NetworkEvents::NetworkEvents()
     : GenericProcessor  ("Network Events")
     , Thread            ("NetworkThread")
-    , threshold         (200.0)
-    , bufferZone        (5.0f)
-    , state             (false)
+    , restart           (false)
+    , changeResponder   (false)
+    , urlport           (0)
 {
     setProcessorType (PROCESSOR_TYPE_SOURCE);
 
-    createZmqContext();
-
-    firstTime = true;
-    responder = nullptr;
-    urlport = 5556;
-    threadRunning = false;
+    if (!setNewListeningPort(5556))
+    {
+        // resort to choosing a port automatically
+        setNewListeningPort(0);
+    }
 
-    opensocket();
+    startThread();
 
     sendSampleCount = false; // disable updating the continuous buffer sample counts,
     // since this processor only sends events
-    shutdown = false;
 }
 
 
-void NetworkEvents::setNewListeningPort (int port)
+bool NetworkEvents::setNewListeningPort(uint16 port)
 {
-    // first, close existing thread.
-    closesocket();
-
-    // allow some time for thread to quit
-#ifdef WIN32
-    Sleep (300);
-#else
-    usleep (300 * 1000);
-#endif
+    ScopedPointer<Responder> newResponder = Responder::makeResponder(port);
```

Comment:
We could make the socket within the thread instead (discarding it without destroying the old one if it fails), and just set urlport here (which we could make atomic) to indicate a port change. Then we wouldn't need `nextResponder` or the lock. I did it this way so that the caller would immediately know whether the change was successful, but that's actually not a big deal since the thread can just update the port on the editor if the switch fails.